src/core/project/qgsproject.h

```diff
@@ -398,6 +400,35 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      */
     void setEllipsoid( const QString &ellipsoid );
 
+    /**
+     * Returns the project's vertical coordinate reference system.
+     *
+     * If the project crs() is a compound CRS, then the CRS returned will
+     * be the vertical component of crs(). Otherwise it will be the value
+     * explicitly set by a call to setVerticalCrs().
+     *
+     * The returned CRS will be invalid if the project has no vertical CRS.
+     *
+     * \see crs()
+     * \see setVerticalCrs()
+     *
+     * \since QGIS 3.38
+     */
+    QgsCoordinateReferenceSystem verticalCrs() const;
+
+    /**
+     * Sets the project's vertical coordinate reference system.
+     *
+     * \note If the project crs() is a compound CRS, then the CRS returned for
+     * verticalCrs() will be the vertical component of crs(). Otherwise it will be the value
```

Comment:
that behavior of ignoring silently setVerticalCrs() when project crs() is a compound is a bit surprising.
I'd rather see one of the 2 below behavior as more expected:
1) emit an exception / error code if calling setVerticalCrs() and that it will be ignored
2) or allow overridding the vertical part of the compound CRS
I cannot really give a geodetic sense to 2), so 1) would seem to be more logical
That said, I don't make this a strong point, given that the currently implemented behavior is well documented

Reply:
@rouault very valid points! How about if I make setVerticalCrs return a bool and an error message string, and explicitly return false with a descriptive error when the compound crs condition is hit?

Reply:
> How about if I make setVerticalCrs return a bool and an error message string, and explicitly return false with a descriptive error when the compound crs condition is hit?

sounds good. You'll perhaps to also have to do that only if the passed vertical CRS != vertical CRS of the projet compound CRS (that is allow to call setVerticalCrs with the vertical CRS of the compound)

Reply:
Ok, done